pyanaconda/modules/subscription/subscription.py

```diff
@@ -174,3 +254,108 @@ def set_system_purpose_data(self, system_purpose_data):
+            raise InvalidValueError(
+                "Invalid subscription request type set '{}'".format(request_type)
+            )
+
+    @property
+    def subscription_request(self):
+        """Subscription request.
+
+        A DBus structure holding data to be used to subscribe the system.
+
+        :return: subscription request DBus structure
+        :rtype: DBusData instance
+        """
+        # Return a deep copy of the subscription request that
+        # has also been cleared of private data.
+        # Thankfully the secret Dbus structures modules
+        # has the get_public_copy() method that does just
+        # that. It creates a deep copy & clears
+        # all SecretData and SecretDataList instances.
+        return get_public_copy(self._subscription_request)
+
+    def set_subscription_request(self, subscription_request):
+        """Set a subscription request.
+
+        Set the complete DBus structure containing subscription
+        request data.
+
+        :param subscription_request: subscription request structure to be set
+        :type subscription_request: DBus structure
+        """
+        self._replace_current_subscription_request(subscription_request)
+        self.subscription_request_changed.emit()
+        log.debug("A subscription request set: %s", str(self._subscription_request))
+
+    def _replace_current_subscription_request(self, new_request):
+        """Replace current subscription request without loosing sensitive data.
+
+        We need to do this to prevent blank SecretData & SecretDataList instances
+        from wiping out previously set secret data. The instances will be blank
+        every time a SubscriptionRequest that went through get_public_copy() comes
+        back with the secret data fields unchanged.
+
+        So what we do is depends on type of the incoming secret data:
+
+        - SECRET_TYPE_NONE - use structure from new request unchanged,
+                             clearing previously set data (if any)
+        - SECRET_TYPE_HIDDEN - secret data has been set previously and
+                               cleared when SubscriptionRequest was sent out;
+                               put secret data from current request to the
+                               new one to prevent it from being lost
+                               (this will also switch the secret data
+                                instance to SECRET_TYPE_TEXT so that
+                                the Subscription module can read it
+                                internally)
+        - SECRET_TYPE_TEXT - this is new secret entry, we can keep it as is
+        """
+        current_request = self._subscription_request
+
+        # Red Hat account password
+        new_hidden = new_request.account_password.type == SECRET_TYPE_HIDDEN
+        current_set = current_request.account_password.type == SECRET_TYPE_TEXT
+        if new_hidden and current_set:
```

Comment:
I think you want to always replace the hidden values. If the new password is hidden and the current one is unset, the new password will stay hidden. It doesn't seem to be right.

Reply:
Yup, always using old value if new one is SECRET_TYPE_HIDDEN is simpler and works. :)